controllers/node_label_module_version_reconciler.go

```diff
@@ -0,0 +1,257 @@
+		return ctrl.Result{}, fmt.Errorf("could not label node %s with added/deleted labels: %v", node.Name, err)
+	}
+
+	return ctrl.Result{Requeue: reconLabelsRes.requeue}, nil
+}
+
+//go:generate mockgen -source=node_label_module_version_reconciler.go -package=controllers -destination=mock_node_label_module_version_reconciler.go nodeLabelModuleVersionHelperAPI
+
+type nodeLabelModuleVersionHelperAPI interface {
+	getLabelsPerModules(nodeLabels map[string]string) map[string]*modulesVersionLabels
+	getModuleLoaderPods(ctx context.Context, nodeName string) ([]v1.Pod, error)
+	reconcileLabels(modulesLabels map[string]*modulesVersionLabels, moduleLoaderPods []v1.Pod) *reconcileLabelsResult
+	updateNodeLabels(ctx context.Context, nodeName string, reconLabelsRes *reconcileLabelsResult) error
+}
+
+type nodeLabelModuleVersionHelper struct {
+	client client.Client
+}
+
+func newNodeLabelModuleVersionHelper(client client.Client) nodeLabelModuleVersionHelperAPI {
+	return &nodeLabelModuleVersionHelper{client: client}
+}
+
+func (nlmvha *nodeLabelModuleVersionHelper) getLabelsPerModules(nodeLabels map[string]string) map[string]*modulesVersionLabels {
+	labelsPerModule := map[string]*modulesVersionLabels{}
+	for key, value := range nodeLabels {
+		if utils.IsVersionLabel(key) {
+			namespace, name, err := utils.GetNamespaceNameFromVersionLabel(key)
+			if err != nil {
+				continue
+			}
+			mapKey := namespace + "-" + name
+			if labelsPerModule[mapKey] == nil {
+				labelsPerModule[mapKey] = &modulesVersionLabels{name: name, namespace: namespace}
+			}
+
+			if utils.IsModuleVersionLabel(key) {
+				labelsPerModule[mapKey].moduleVersionLabel = value
+			}
+			if utils.IsModuleLoaderVersionLabel(key) {
+				labelsPerModule[mapKey].moduleLoaderVersionLabel = value
+			}
+			if utils.IsDevicePluginVersionLabel(key) {
+				labelsPerModule[mapKey].devicePluginVersionLabel = value
+			}
+		}
+	}
+
+	return labelsPerModule
+}
+
+func (nlmvha *nodeLabelModuleVersionHelper) getModuleLoaderPods(ctx context.Context, nodeName string) ([]v1.Pod, error) {
+	var modulePodsList v1.PodList
+	fieldSelector := client.MatchingFields{"spec.nodeName": nodeName}
+	labelSelector := client.MatchingLabels{constants.DaemonSetRole: constants.ModuleLoaderRoleLabelValue}
+	err := nlmvha.client.List(ctx, &modulePodsList, labelSelector, fieldSelector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get list of all module loader pods for on node %s: %v", nodeName, err)
+	}
+	return modulePodsList.Items, nil
+}
+
+func (nlmvha *nodeLabelModuleVersionHelper) updateNodeLabels(ctx context.Context, nodeName string, reconLabelsRes *reconcileLabelsResult) error {
+	node := v1.Node{}
+
+	if err := nlmvha.client.Get(ctx, types.NamespacedName{Name: nodeName}, &node); err != nil {
+		return fmt.Errorf("could not get node %s: %v", nodeName, err)
+	}
+
+	nodeCopy := node.DeepCopy()
+
+	if node.Labels == nil && len(reconLabelsRes.labelsToAdd) > 0 {
+		node.Labels = make(map[string]string, 1)
```

Comment:
```suggestion
		node.Labels = map[string]string{}
```

I don't see the point of setting the initial map size to 1 when we don't know how many labels we are going to add. There is no logical reason to set the value to 1.

Am I missing something?

Reply:
fixed